repo/repo.go

```diff
@@ -167,13 +223,33 @@ func FindExternalRepo(repoRoot, name string) (string, error) {
 }
 
 // ListRepositories extracts metadata about repositories declared in a
-// WORKSPACE file.
-//
-// The set of repositories returned is necessarily incomplete, since we don't
-// evaluate the file, and repositories may be declared in macros in other files.
-func ListRepositories(workspace *rule.File) []Repo {
-	var repos []Repo
-	for _, r := range workspace.Rules {
+// file.
+func ListRepositories(workspace *rule.File) (repos []Repo, repoNamesByFile map[*rule.File][]string, err error) {
+	repoNamesByFile = make(map[*rule.File][]string)
+	repos, repoNamesByFile[workspace] = getRepos(workspace.Rules)
+
+	for _, d := range workspace.Directives {
+		switch d.Key {
+		case "repository_macro":
+			vals := strings.Split(d.Value, "%")
+			if len(vals) != 2 {
+				return nil, nil, fmt.Errorf("Failure parsing repository_macro: %s, expected format is macroFile%%defName", d.Value)
+			}
+			macroFile, err := rule.LoadMacroFile(vals[0], "", vals[1])
```

Comment:
The macro file name should be a relative path from `filepath.Dir(workspace.Path)`.

Make sure to run `filepath.Clean` on it before joining it, and make sure it doesn't start with `".."`.